Add RSV log line

valarnin opened the pull request:
Completely untested but should, in theory, work. I should be able to test this when one of my groups runs later tonight.

Potential issues:
1. As it currently stands this might produce false positives if there's uncleared memory between large packets?
2. Technically requires 4 byte comparisons, rather than 1 for the other log lines
3. Probably not very resilient to version changes, but considering the other added log lines will require opcode updates anyways, it's probably okay?

xpdota commented:
Is there an example of what the lines look like?

valarnin commented:
> Is there an example of what the lines look like?

I haven't zoned into any instance that sends these lines to actually be able to provide a full example, but here's a reconstruction based on previous data:

```
260|2022-08-30T20:12:47.2390000+02:00|12000000|_rsv_30478_-1_2_0_0_SE2DC5B04_EE2DC5B04|Kralle und Schwanz|0123456789ABCDEF
```

It may make sense to include language as well I guess? At which point maybe it should be:


```
260|2022-08-30T20:12:47.2390000+02:00|de|12000000|_rsv_30478_-1_2_0_0_SE2DC5B04_EE2DC5B04|Kralle und Schwanz|0123456789ABCDEF
```

tssailzz8 commented:
https://github.com/tssailzz8/RSVfinder
i think we can use dalamudPlugins to do it,it will be fast than use act

Akurosia commented:
Hi, is there any plants landing this before omega ultimate happens?

xpdota commented:
I would also like to have something like this.

RSV stuff for current content is "removed" in the sense that the ACT plugin has the real names overridden, but is still present in the sense that the game files (and by extension, xivapi and friends) have the _rsv names.

valarnin commented:
Their presence in the game client is irrelevant, the question is "does the network packet still get sent on zone in"?

quisquous commented:
> I ended up shelving this due to discussion in the Discord about it potentially being put into the base FFXIV_ACT_Plugin. If that's not the case anymore, then I can rebase it and mark it as ready for review.

Sorry for the distraction of that conversation.  It felt like a logically better place to do it (then only the ffxiv plugin has to do it, and fflogs and other downstream consumers of logs don't have to care about these), but it didn't sound like Ravahn was interested at all in handling it.

valarnin commented:
I have rebased this and it's ready for review.

I think there's still an outstanding question of "should we be using blind matching heuristics on these lines instead of matching an opcode?"

I don't know if there's a fast or solo way to get the RSV opcode like there is for MapEffect and CEDirector, which is why I opted for the heuristic approach (packet size is big enough and `_rsv` characters are in the right spot in the packet).

valarnin commented:
Alright, that should be the outstanding issues resolved, but as I expected there's no `_rsv` line transmitted for p5s-p8s anymore, so I don't have a good way to test it.

I'm pretty sure patch 6.30 removed _rsv lines from DSR as well.

xpdota commented:
That is correct, there are currently no RSV things in the game in the Intl client. CN/KR might be the only way to test, other than waiting for TOP.

Makar8000 commented:
> I don't know if there's a fast or solo way to get the RSV opcode like there is for MapEffect and CEDirector, which is why I opted for the heuristic approach (packet size is big enough and `_rsv` characters are in the right spot in the packet).

It should be simple enough to get it through static analysis, which is what I'm currently doing for the MapEffect and CEDirector PRs.  Otherwise, the only way is to queue into the content itself that provides it. 
Not suggesting one way or the other, just pointing it out incase there's a clear preference.

quisquous commented:
I don't feel strongly, re: opcode vs heuristic.  If it's easy to pull this out with static analysis (I am not sure I know how to do that, so pointers would be helpful too?), then maybe using a real opcode is better, especially if opcodes update immediately.  Looking for the rsv string in data seems a little sketchy (maybe false positives, maybe they stop naming abilities that way), but I am probably overthinking it.

Makar8000 commented:
> I don't feel strongly, re: opcode vs heuristic. If it's easy to pull this out with static analysis (I am not sure I know how to do that, so pointers would be helpful too?), then maybe using a real opcode is better, especially if opcodes update immediately. Looking for the rsv string in data seems a little sketchy (maybe false positives, maybe they stop naming abilities that way), but I am probably overthinking it.

I can share the IDA script I use. It's a slightly modified version of Adam's that he goes over [here](https://sapphireserver.github.io/dev/2019/12/23/fixing-opcodes.html). This potentially has it's own problems though, and I'm not sure what would be more consistent honestly. At least for opcodes, even if the script fails, it's is simple enough to check in-game and wouldn't require an OP update to resolve. It's a pick-your-poison scenario I think.

valarnin commented:
I feel that since Makar is the one currently doing opcode PRs for updates for CEDirector/MapEffect, they're probably in the best position to determine if using an opcode for this going forward is a better option. I'm fine with either option, and I'm fine with rewriting the PR to use opcode if that's the preferred method.

Makar8000 commented:
Truthfully, I only learned to do this since 6.1, so I'm still not very experienced with it to be honest. It may be best to consult one of the Sapphire folks on their opinion. I just know my current method with opcodes hasn't failed me yet 😅

xpdota commented:
Why not both? Opcode if available, otherwise fall back to heuristic? That would give you RSV data immediately on plugin update, while still being able to use the cleaner option after an opcode update.

quisquous commented:
> Why not both? Opcode if available, otherwise fall back to heuristic? That would give you RSV data immediately on plugin update, while still being able to use the cleaner option after an opcode update.

Sorry, disagree.  Both = more complication and harder to debug when it's failing.  We should pick one path.

valarnin commented:
Does anyone have the corresponding debug network 252 lines for TOP available for me to test with? Just an ability and a status (with description) should be sufficient, but if you've got all the corresponding 252 lines that would be best.

Akurosia commented:
@valarnin english one is here: https://discord.com/channels/551474815727304704/551527590800654386/1067389306546704405
All other languages is here: https://discord.com/channels/551474815727304704/551527590800654386/1067560599036362853

valarnin commented:
@Akurosia None of those messages and none of the log files in the contribute-logfiles channel have the ***raw 252*** log line, with the raw byte data I would need to validate the packet structure.

Akurosia commented:
> @Akurosia None of those messages and none of the log files in the contribute-logfiles channel have the _**raw 252**_ log line, with the raw byte data I would need to validate the packet structure.

sry my bad i can send you the raw jp/de/fr once as i generated them myself, will send you these files on discord

Makar8000 commented:
> Does anyone have the corresponding debug network 252 lines for TOP available for me to test with? Just an ability and a status (with description) should be sufficient, but if you've got all the corresponding 252 lines that would be best.

Since you the raw EN one in DM

valarnin commented:
Rebased to main due to some weirdness on my end, and switched the line to use opcode instead.

Pretty sure the opcode is correct but if someone wants to double check that would be appreciated. Also I still can't test this fully because it's going to be a while before my group is ready to step into TOP, so if someone wants to test this and verify that it's emitting lines as expected on zone in that would help.

xpdota commented:
I would also be interested in a sample log. I'll be in tomorrow if nobody else posts a log between now and then.

Makar8000 commented:
Has any testing been done with this? Curious on the progress of this PR.
FYI, there is a hotfix maintenance in about 48 hours, so the opcode will probably need an update soon.

Makar8000 commented:
For `2023.02.03.0000.0000`, the RSV opcode is `375` base10

valarnin commented:
Hmm, interesting. FFXIVClientStructs now has this struct:

```cs
[StructLayout(LayoutKind.Explicit, Size = 0x228)]
public unsafe partial struct LayoutWorld
{
    [StaticAddress("48 89 05 ?? ?? ?? ?? 48 8B 00", 3, isPointer: true)]
    public static partial LayoutWorld* Instance();

    [FieldOffset(0x20)] public LayoutManager* ActiveLayout;
    [FieldOffset(0x218)] public StdMap<Utf8String, Pointer<byte>>* RsvMap;
}
```

valarnin commented:
Alright, rebased to main, fixed the now-conflicting line ID, fixed issue with CRLF, and added latest global opcode.

If you plan on consuming these lines, you'll need to replace contents such that `\r` and `\n` get unescaped to their correct raw values. There's probably a better way to handle these going forward but I'm not sure what that would be off the top of my head.

This is working as expected from my testing, see attached file for RSV lines from zoning into TOP. Note that these lines were generated with the previous line ID of 261, the new line ID is 262.

[rsv lines.txt](https://github.com/OverlayPlugin/OverlayPlugin/files/11300239/rsv.lines.txt)

This is now ready for review/merge.
